engineering/architecture/dateandtime.md

```diff
@@ -0,0 +1,107 @@
+
+Another way to input data is through bulk import. In that case, we're completely reliant on the database parsing. This is the reason goal (4) is important.
+
+#### Regarding relative times
+
+There are some relative dates / times which make sense, e.g. "tomorrow", to store in a date column. These will _not_ be stored in a "relative" way, though. They'll be stored as a static date or time (one day later than the current day in the example). As for durations, there's some weird overlap between intervals "2 days ago" and a specific time "2 days ago". We'll choose whether to store the duration or a static time based on the column type.
+
+## Overall output flow
+
+For showing dates / times, they'll always be returned from the API in the appropriate canonical form described below. It's then up to the UI layer to format them for a pleasant display to the user.
+
+## RFC 3339 (based on ISO 8601)
+
+The canonical representation of a date, time, datetime, or duration in Mathesar will be as per the [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339#appendix-A) spec, with very minor modifications to match our use case.
+
+## Canonical Representation of Durations
+
+A duration describes the amount of time between two datetime values (or dates or times). Two durations can be added or subtracted to create another duration. A duration can be added to a date, time, or datetime to get a different date, time, or datetime. As per the RFC, the canonical representation of a duration is:
+
+```python
+f'P{years}Y{months}M{days}DT{hours}H{minutes}M{seconds}S'
+```
+Here, `seconds` can be an integer or float, and the other variables are all integers. For the canonical representation, we are filling in each unit with zeroes, which is optional in the RFC spec. To be completely proper, we aggregate units of time into larger ones where possible:
+
+- seconds aggregate into minutes,
+- minutes aggregate into hours, and
+- months aggregate into years.
+
+Because some units of time (months and days) are inconsistent w.r.t. the number of smaller units they contain, we avoid aggregating some units:
+
+- hours do not aggregate into days, and
+- days do not aggregate into months.
+
+#### Examples of Canonical Durations
+
+- `'1 year 1 month, 1 day 1 hour 1 minute 1.1 seconds'` maps to `P1Y1M1DT1H1M1.1S`
+- `'0 years 13 months 31 days 23 hours 60 minutes 61.1 seconds'` maps to `P1Y1M31DT24H1M1.1S`
+
+Note where the units do and do not aggregate into larger units in the second example.
+
+## Canonical Representation of Dates and Times
+
+These will similarly follow RFC 3339.
+
+### Dates
+
+```python
+f'{year}-{month}-{day} {era}'
+```
+Here, `year` will have a minimum of 4 digits, using preceding zeroes for years like 0042 b.c. Years with more than 4 digits have the correct number of digits. `month` is a 2-digit integer, as is `day` (with preceding zeroes as needed). `era` will always be included, and will be either `BC` or `AD` as appropriate.
```

Comment:
> Here, `year` will have a minimum of 4 digits

I'm happy to see that we'll be [Y10k](https://datatracker.ietf.org/doc/html/rfc2550) compliant 😁